### PR TITLE
Support more terminal attributes other than bold.

### PR DESCRIPTION
**Pull request by MaskRay**
And remove some redundant bold definitions in r_cons.h

The macros in header files should be reduced.

https://github.com/tmux/tmux/wiki/FAQ#i-dont-see-italics-or-italics-and-reverse-are-the-wrong-way-round

Finally got a chance to cite this https://gist.github.com/XVilka/8346728

### CODE REVIEW

File: libr/cons/pal.c [hunk 1]
```diff
@@ -118,7 +111,8 @@ R_API void r_cons_pal_init () {
 	cons->cpal.args               = (RColor) RColor_YELLOW;
 	cons->cpal.bin                = (RColor) RColor_CYAN;
 	cons->cpal.btext              = (RColor) RColor_YELLOW;
-	cons->cpal.call               = (RColor) RColor_BGREEN;
+	cons->cpal.call               = (RColor) RColor_GREEN;
+	cons->cpal.call.attr          = 1u << 1;
```
Comment (xarkes) [inline libr/cons/pal.c]: Why don't you use a `#define`? Something like RCOLOR_BOLD or RCOLOR_ATTR_BOLD?

File: libr/cons/pal.c [hunk 2]
```diff
@@ -132,7 +126,8 @@ R_API void r_cons_pal_init () {
 	cons->cpal.fname              = (RColor) RColor_RED;
 	cons->cpal.help               = (RColor) RColor_GREEN;
 	cons->cpal.input              = (RColor) RColor_WHITE;
-	cons->cpal.invalid            = (RColor) RColor_BRED;
+	cons->cpal.invalid            = (RColor) RColor_RED;
+	cons->cpal.invalid.attr       = 1u << 1;
```
Comment (xarkes) [inline libr/cons/pal.c]: Why don't you keep BRED? Saves one line.

File: libr/cons/pal.c (hunk 1 above)
Comment (xarkes) [reply libr/cons/pal.c]: Or an enum

File: libr/include/r_cons.h [hunk 3]
```diff
@@ -503,39 +505,38 @@ typedef struct r_cons_t {
...
+#define RColor_YELLOW    RCOLOR(0      , ALPHA_FG, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00)
+#define RColor_BGYELLOW  RCOLOR(0      , ALPHA_BG, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00)
+#define RColor_CYAN      RCOLOR(0      , ALPHA_FG, 0x00, 0xff, 0xff, 0x00, 0x00, 0x00)
+#define RColor_BGCYAN    RCOLOR(0      , ALPHA_BG, 0x00, 0xff, 0xff, 0x00, 0x00, 0x00)
+#define RColor_BLUE      RCOLOR(0      , ALPHA_FG, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00)
+#define RColor_BGBLUE    RCOLOR(0      , ALPHA_BG, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00)
+#define RColor_GRAY      RCOLOR(0      , ALPHA_FG, 0x7f, 0x7f, 0x7f, 0x00, 0x00, 0x00)
+#define RColor_BGGRAY    RCOLOR(0      , ALPHA_BG, 0x7f, 0x7f, 0x7f, 0x00, 0x00, 0x00)
+#define RColor_BBLACK    RCOLOR(1u << 1, ALPHA_FG, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
+#define RColor_BRED      RCOLOR(1u << 1, ALPHA_FG, 0xff, 0x00, 0x00, 0x00, 0x00, 0x00)
```
Comment (xarkes) [inline libr/include/r_cons.h]: Keep or delete, but if you keep it use it: https://github.com/radare/radare2/pull/9535/files#diff-956422fb542ea7cb490dda6e154b37dbR130

File: libr/cons/pal.c (hunk 1 above)
Comment (MaskRay) [reply libr/cons/pal.c]: It is not a color so I don't use `RCOLOR_*`

File: libr/cons/pal.c (hunk 2 above)
Comment (MaskRay) [reply libr/cons/pal.c]: macros are bad and should not pollute header files.

File: libr/cons/pal.c (hunk 1 above)
Comment (xarkes) [reply libr/cons/pal.c]: Yes but imho having an enum is better than having to guess what 1u << 3 means.

File: libr/cons/pal.c (hunk 2 above)
Comment (xarkes) [reply libr/cons/pal.c]: Yes indeed, see my comment below then.

File: libr/include/r_cons.h (hunk 3 above)
Comment (MaskRay) [reply libr/include/r_cons.h]: Removed a few bold macros

File: libr/include/r_cons.h [hunk 4]
```diff
@@ -142,9 +142,12 @@ enum {
 
 enum { COLOR_MODE_DISABLED = 0, COLOR_MODE_16, COLOR_MODE_256, COLOR_MODE_16M };
 
-enum { ALPHA_RESET = 0x00, ALPHA_FG = 0x01, ALPHA_BG = 0x02, ALPHA_FGBG = 0x03, ALPHA_BOLD = 0x04 };
+enum { ALPHA_RESET = 0x00, ALPHA_FG = 0x01, ALPHA_BG = 0x02, ALPHA_FGBG = 0x03 };
+enum { TERM_ATTR_BOLD = 1 << 1 };
```
Comment (xarkes) [inline libr/include/r_cons.h]: Why an enum if you have only one value? What about TERM_ATTR_ITALIC, TERM_ATTR_UNDERLINE, ...?
Haha

Comment (MaskRay) [reply libr/include/r_cons.h]: TERM_ATTR_ITALIC ... are not used much so I do not create an enum constant for them.

Comment (xarkes) [reply libr/include/r_cons.h]: Meh ok

**Comment by xarkes** [comment]
Lgtm

File: libr/cons/pal.c [hunk 5]
```diff
@@ -298,6 +300,30 @@ R_API char *r_cons_pal_parse (const char *str, RColor *outcol) {
 			}
 		}
 	}
+	if (attr) {
+		// Parse extra attributes.
+		const char *p = attr;
+		while (p) {
+			if (!memcmp(p, "bold", 4)) {
```
Comment (radare) [inline libr/cons/pal.c]: use strncmp or r_str_ncmp but memcmp on strings can result on oobreads

Comment (radare) [reply libr/cons/pal.c]: add missing space before every (

File: libr/cons/rgb.c [hunk 6]
```diff
@@ -164,15 +164,28 @@ R_API char *r_cons_rgb_str_off(char *outstr, ut64 off) {
...
+	outstr[1] = '[';
+	for (; attr; attr &= attr - 1) {
+		switch (attr & -attr) {
+		case 1u << 1: outstr[i] = '1'; break;
+		case 1u << 2: outstr[i] = '2'; break;
+		case 1u << 3: outstr[i] = '3'; break;
+		case 1u << 4: outstr[i] = '4'; break;
+		case 1u << 5: outstr[i] = '5'; break;
+		}
+		outstr[i+1] = ';';
```
Comment (radare) [inline libr/cons/rgb.c]: spaces around +

File: libr/include/r_cons.h (hunk 4 above)
Comment (radare) [reply libr/include/r_cons.h]: yeah will be good to have them prefixed by RColor or R_CONS_ATTR_*

**Comment by radare** [comment]
What about windows and HTML? i think we should tweak the ansi parsers to support them, so the webui and windows users may benefit out of this.

File: libr/cons/rgb.c (hunk 6 above)
Comment (MaskRay) [reply libr/cons/rgb.c]: Done

**Comment by MaskRay** [comment]
No idea about windows or HTML. Renamed to R_CONS_ATTR_BOLD but not sure how many variants should be added to that enum

**Pull request merged**